certora/specs/RecoveryConfirmationSignatureValidity.spec

```
@@ -0,0 +1,142 @@
+}
+
+// This rule checks that the number of approvals counted by the contract is equal to the number of valid signatures.
+rule approvalsCountShouldEqualTheAmountOfSignatures(env e) {
+    mathint recoveryConfirmationCountBefore = recoveryConfirmationCount;
+    address _wallet;
+    address[] _newOwners;
+    uint256 _newThreshold;
+    uint256 walletNonce = currentContract.nonce(_wallet);
+    SocialRecoveryModule.SignatureData[] signatures;
+    bool _execute;
+    bytes32 recoveryHash = getRecoveryHash(
+        e,
+        _wallet,
+        _newOwners,
+        _newThreshold,
+        walletNonce
+    );
+
+    multiConfirmRecovery(e, _wallet, _newOwners, _newThreshold, signatures, _execute);
+
+    assert to_mathint(signatures.length) == recoveryConfirmationCount - recoveryConfirmationCountBefore, "More approvals counted than valid signatures";
+}
+
+// This rule checks that only supplied signatures and signers are counted as approvals.
+rule noShadowApprovals(env e) {
+    address _wallet;
+    address[] _newOwners;
+    uint256 _newThreshold;
+    uint256 walletNonce = currentContract.nonce(_wallet);
+    SocialRecoveryModule.SignatureData[] signatures;
+    bool _execute;
+    bytes32 recoveryHash = getRecoveryHash(
+        e,
+        _wallet,
+        _newOwners,
+        _newThreshold,
+        walletNonce
+    );
+    address otherAddress;
+    // We need to correctly initialize pre-state to ensure that the `otherAddress` is not a signer and has not confirmed the recovery.
+    require !currentContract.isGuardian(_wallet, otherAddress) && !currentContract.confirmedHashes[recoveryHash][otherAddress];
+
+    multiConfirmRecovery(e, _wallet, _newOwners, _newThreshold, signatures, _execute);
+
+    assert forall uint256 i. i < signatures.length => currentContract.confirmedHashes[recoveryHash][signatures[i].signer], "Approvals were not correctly set";
+    assert !currentContract.confirmedHashes[recoveryHash][otherAddress], "Other address should not be able to confirm recovery";
+}
+
+// This rule verifies that specifying the same signer twice in the signatures array will cause the transaction to revert.
+// We cannot really verify the actual signature duplication because we summarize the `isValidSignatureNow` function, which makes the run time out if not summarized.
+rule duplicateSignersRevert(env e) {
+    address _wallet;
+    address[] _newOwners;
+    uint256 _newThreshold;
+    uint256 walletNonce = currentContract.nonce(_wallet);
+    SocialRecoveryModule.SignatureData[] signatures;
+    bool _execute;
+    bytes32 recoveryHash = getRecoveryHash(
+        e,
+        _wallet,
+        _newOwners,
+        _newThreshold,
+        walletNonce
+    );
+    uint256 i1; uint256 i2;
+    require i1 != i2;
+    require signatures[i1].signer == signatures[i2].signer;
+
+    multiConfirmRecovery@withrevert(e, _wallet, _newOwners, _newThreshold, signatures, _execute);
+
+    assert lastReverted, "Duplicate signers should revert";
```

Ubernit: Certora recommends the following pattern to prevent accidentally clobbering the `lastReverted` variable. It could happen if you, for example, have a call to a view method in the assert and move the conditions around:


```suggestion
    bool multiConfirmReverted = lastReverted;

    assert multiConfirmReverted, "Duplicate signers should revert";
```